### ISSUE TITLE
Proposal: Compile-time checking for Bazel Go rules

### ISSUE THREAD

**Issue by stjj89** [bazelbuild/rules_go#1097]
# Proposal: Compile-time checking for Bazel Go rules

## Overview

Bazel currently supports Java compile-time checking via [Error Prone](https://blog.bazel.build/2015/06/25/ErrorProne.html). This allows Bazel to check for common mistakes in Java code that will not be caught by the compiler, and fail builds of erroneous code, if necessary.

Internally, we have a similar compile-time checking framework for Go called nogo, which we would like to add to Bazel. nogo will run alongside Bazel builds and fail builds that do not pass its checks. Users will be able to turn off nogo or configure its checks via Bazel builds.

## Detailed design

nogo will be implemented as a library that runs in parallel to the Go compiler in the [compile program](https://github.com/bazelbuild/rules_go/blob/master/go/tools/builders/compile.go). It will include several correctness and security checks that:
* Have an extremely low false-positive rate (~1%), and
* Emit an informative error message that will make the issue in the code obvious or offer a simple fix.

### Correctness checks
Correctness checks will ensure a correctness property in Go programs. Examples include [a check for goroutines sharing a loop iterator variable](https://github.com/golang/go/wiki/CommonMistakes#using-goroutines-on-loop-iterator-variables), and ports of the [atomic](https://golang.org/cmd/vet/#hdr-Atomic_mistakes), [bool](https://golang.org/cmd/vet/#hdr-Boolean_conditions), [buildtags](https://golang.org/cmd/vet/#hdr-Build_tags), [nilfunc](https://golang.org/cmd/vet/#hdr-Nil_function_comparison), and [printf](https://golang.org/cmd/vet/#hdr-Printf_family) vet checks that the Go team has deemed [100% correct](https://github.com/golang/go/commit/0d188752524282496ebd0ab4b382bb4ff8750c90#diff-e83f78900998667ec15c83c5b59ba017). 

### Security checks
Security checks will ensure security properties in Go programs that are subtle and require considerable domain expertise to reason about. These checks will be highly configurable to allow Bazel users to customize them to fit their own security policies. Examples of these checks include checks for imports or calls to dangerous or security-sensitive functions or packages.

### Configuring checks
nogo checks that are configurable will be configured via a JSON file. The configuration file will have the following format:

```json
{
  "disallowedpackages": [
    {
      "pkg": "crypto/des",
      "whitelistpkg": "my/project/whitelist/des",
    },
    {
      "pkg": "crypto/subtle",
      "whitelistpkg": "my/project/whitelist/subtle",
    },
  ],
  "disallowedfunctions": [
    {
      "func": "some/pkg/deprecated.Func",
      "whitelistpkg": "my/project/whitelist/deprecatedFunc",
    },
    {
      "func": "some/pkg/(*dangerous).Method",
      "whitelistpkg": "my/project/whitelist/dangerousMethod",
    },
  ]
}
```

Each top-level key in this map will name a configurable check. Each check can define its own format for its configuration data.

Bazel users will be able to disable nogo entire as well, either via a build flag (e.g. `--disable_go_checks`) or an entry in the configuration file (e.g. `"disable: true"`).

### vet running with `go test`

In Go 1.10, vet will [run alongside `go test`](https://groups.google.com/forum/#!msg/golang-dev/qfa3mHN4ZPA/X2UzjNV1BAAJ) and cause a test failure if its checks do not pass. Only a subset of tests that are 100% correct will run during `go test`. If vet checks will also fail calls to `bazel test` for Go builds, it might make nogo seem redundant. Potentially, both vet (during `go test`) and nogo can co-exist, and we will distinguish between checks that cause build failures and checks that cause test failures.

### Specifying the configuration file

I'm not sure what is the best way to specify the nogo configuration file. Should it be passed via a function invoked in the `WORKSPACE` file? e.g.

```python {WORKSPACE}
...
load("@io_bazel_rules_go//go:def.bzl", "go_rules_dependencies", "go_register_toolchains")
go_rules_dependencies()
go_register_toolchains()
nogo_config_file("path/to/nogo.config")
```

### Pluggable checks

In the future, I would like nogo to support custom pluggable checks, just like [Error Prone does](http://errorprone.info/docs/plugins). However, supporting such checks will probably involve significant work to enable dynamic loading of Go code in Bazel. If and when this work is finally done, we should be able to add this feature to nogo without changing this proposed interface (e.g. by adding a new `nogo_plugin` attribute to Go rules).

**Comment by stjj89** [comment]
Closing this since [`nogo`](https://github.com/bazelbuild/rules_go/blob/master/go/nogo.rst) has made it into the latest release of the Go rules.
